### ISSUE TITLE
BatchNormalization test fails with invalid memory access 

### ISSUE THREAD

**Issue by npanpaliya** [soumith/cudnn.torch#255]
On ppc64le, Ubuntu 16.04/14.04, cuda 8.0/7.5, we are testing torch/extra/cudnn/test/test.lua with cudnn 5.1. From test.lua, we are getting a test failed for BatchNormalization with below error. 
**Error:**
THCudaCheck FAIL file=/home/ac/pkgbuild/torch/torch/extra/cutorch/lib/THC/generic/THCStorage.cu line=41 error=77 : an illegal memory access was encountered
1/1 BatchNormalization .................................................. [ERROR]
/opt/DL/torch/bin/luajit: cuda runtime error (77) : an illegal memory access was encountered at /home/ac/pkgbuild/torch/torch/extra/cutorch/lib/THC/generic/THCStorage.c:158

Analysis:
- For this test, testBatchNormalization is called with input size having 2 items and in testBatchNormalization accesses inputSize[2] which is also 2nd item. If I either change inputSize[2] to inputSize[1] in testBatchNormalization or I add one more item in the inputSize, then above error goes away (as I'd earlier thought even Lua will have array indexing starting from zero). But even with this, I further get CUDNN_STATUS_BAD_PARAM error.
- The exact call causing above failure of invalid memory access are cbn:forward and cbn:backward.

I also wanted to know how we can debug these tests. I'm new to torch and lua, so need some help to go ahead. Any pointers would be helpful. 
Thanks.

**Comment by ngimel** [comment]
That's a bug somewhere in ppc luajit/ffi when passing double arguments by value. To confirm, print out batch norm arguments as seen by errcheck and as seen by cudnn (actually, not cudnn but a dummy library function that implements the same interface:

1/1 BatchNormalization .................................................. [WAIT]number of args    17    
cdata<struct cudnnContext *>: 0x01001db6c080 <--arguments as passed to errcheck
CUDNN_BATCHNORM_PER_ACTIVATION    
cdata<float *>: 0x01001da8a750
cdata<float *>: 0x01001da8a830
cdata<struct cudnnTensorStruct *>: 0x0100250afbf0
cdata<float *>: 0x230aac0000
cdata<struct cudnnTensorStruct *>: 0x0100250affd0
cdata<float *>: 0x230aac2000
cdata<struct cudnnTensorStruct *>: 0x0100250b0570
cdata<float *>: 0x230a7c0a00
cdata<float *>: 0x230a7c1000
0.1    
cdata<float *>: 0x230a7c0800
cdata<float *>: 0x230a7c1200
0.001    
cdata<float *>: 0x230a7c2600
cdata<float *>: 0x230a7c2800
training, double args passed by value    
alpha 0x1001da8a750  <--arguments as received by library
beta 0x1001da8a830
bottomDesc 0x100250afbf0
bottomData 0x230aac0000
topDesc 0x100250affd0
topData 0x230aac2000
sdesc 0x100250b0570
scale 0x230a7c0a00
bias 0x230a7c1000
ema 0.100000
runningmean 0x230a7c1200 <-- note, everything after the first double argument is mangled. runningMean and runningVar take on values of some different vars, saveMean and saveVar are garbage
runningvar 0x230a7c2600
eps 0.001000
saveMean 0x7
saveVar 0x3fffcdd6d5a8

CUDNN_STATUS_BAD_PARAM is expected when you are changing the arguments.

**Comment by soumith** [comment]
We can recommend all ppc64 users to use Lua 5.2

**Comment by npanpaliya** [comment]
Thanks @ngimel and @soumith for your response. I'll try to update Lua to 5.2 and see if it fixes the problem.

**Comment by npanpaliya** [comment]
I installed lua5.2 and tried to build torch by changing TORCH_LUA_VERSION to LUA52 in install.sh as per this [link](http://torch.ch/docs/getting-started.html). 
But now, on th> require 'cudnn' I get below error -
th> require 'cudnn'
Trace/breakpoint trap

Did I miss any other setting or pre-requisite?

**Comment by ngimel** [comment]
WIth lua 5.2 you also need luaffifb
https://github.com/facebook/luaffifb

**Comment by npanpaliya** [comment]
We've reported the problem to luajit compiler team. Also, found some alternative without updating to lua5.2 that fixes the problem. So, we can close this issue. 

Thanks for your prompt and effective responses.

**Comment by hartb** [comment]
Team member of @npanpaliya here.  Just in case it's useful to anyone in the future...  We found a problem in the way the lua JIT constructs long argument lists for POWER. The problem should be confined to cases where a floating point argument beyond position 8 is followed by non-FP arguments. We'll work to get a fix into to lua JIT.
